File: src/ml/inference/health.py
"""
Health Check Endpoints for ML Service
Provides health, readiness, and metrics endpoints
"""

import time
import psutil
from datetime import datetime
from typing import Dict, Any

from fastapi import APIRouter, Response
from pydantic import BaseModel

router = APIRouter()

# Track start time for uptime calculation
START_TIME = time.time()
REQUEST_COUNT = 0


class HealthResponse(BaseModel):
    """Health check response model"""
    status: str
    timestamp: str
    uptime: int


class ReadinessResponse(BaseModel):
    """Readiness check response model"""
    status: str
    timestamp: str
    checks: Dict[str, str]


class InfoResponse(BaseModel):
    """Application info response model"""
    name: str
    version: str
    environment: str
    python_version: str
    timestamp: str


@router.get("/health", response_model=HealthResponse)
async def health_check():
    """
    Basic health check endpoint.
    Returns 200 if the service is running.
    """
    global REQUEST_COUNT
    REQUEST_COUNT += 1

    return HealthResponse(
        status="healthy",
        timestamp=datetime.utcnow().isoformat() + "Z",
        uptime=int(time.time() - START_TIME)
    )


@router.get("/health/ready", response_model=ReadinessResponse)
async def readiness_check():
    """
    Readiness probe - checks if app can serve traffic.
    Verifies model is loaded and dependencies are available.
    """
    global REQUEST_COUNT
    REQUEST_COUNT += 1

    checks = {
        "model": await check_model_loaded(),
        "memory": check_memory_available(),
        "database": await check_database_connection()
    }

    all_healthy = all(v == "healthy" for v in checks.values())

    return ReadinessResponse(
        status="ready" if all_healthy else "not_ready",
        timestamp=datetime.utcnow().isoformat() + "Z",
        checks=checks
    )


@router.get("/health/live")
async def liveness_check():
    """
    Liveness probe - checks if app is running.
    Simple check that returns immediately.
    """
    global REQUEST_COUNT
    REQUEST_COUNT += 1

    return {"status": "alive", "timestamp": datetime.utcnow().isoformat() + "Z"}


@router.get("/health/metrics")
async def metrics():
    """
    Prometheus-compatible metrics endpoint.
    Returns metrics in Prometheus text format.
    """
    global REQUEST_COUNT
    REQUEST_COUNT += 1

    uptime = int(time.time() - START_TIME)
    memory = psutil.Process().memory_info()
    cpu_percent = psutil.Process().cpu_percent()

    metrics_text = f"""# HELP meal_assistant_ml_uptime_seconds Time since ML service started
# TYPE meal_assistant_ml_uptime_seconds gauge
meal_assistant_ml_uptime_seconds {uptime}

# HELP meal_assistant_ml_request_total Total number of health check requests
# TYPE meal_assistant_ml_request_total counter
meal_assistant_ml_request_total {REQUEST_COUNT}

# HELP meal_assistant_ml_memory_rss_bytes Resident set size in bytes
# TYPE meal_assistant_ml_memory_rss_bytes gauge
meal_assistant_ml_memory_rss_bytes {memory.rss}

# HELP meal_assistant_ml_memory_vms_bytes Virtual memory size in bytes
# TYPE meal_assistant_ml_memory_vms_bytes gauge
meal_assistant_ml_memory_vms_bytes {memory.vms}

# HELP meal_assistant_ml_cpu_percent CPU usage percentage
# TYPE meal_assistant_ml_cpu_percent gauge
meal_assistant_ml_cpu_percent {cpu_percent}

# HELP python_info Python version info
# TYPE python_info gauge
python_info{{version="3.11"}} 1
"""

    return Response(content=metrics_text.strip(), media_type="text/plain")


@router.get("/health/info", response_model=InfoResponse)
async def info():
    """
    Application info endpoint.
    Returns service metadata.
    """
    import sys
    import os

    return InfoResponse(
        name="meal-assistant-ml",
        version="1.0.0",
        environment=os.getenv("ENVIRONMENT", "development"),
        python_version=sys.version.split()[0],
        timestamp=datetime.utcnow().isoformat() + "Z"
    )


# =============================================================================
# Health Check Helper Functions
# =============================================================================

async def check_model_loaded() -> str:
    """Check if ML models are loaded and ready."""
    try:
        # In production, this would check if models are actually loaded
        # Example: if model_instance is not None and model_instance.is_ready()
        return "healthy"
    except Exception as e:
        return f"unhealthy: {str(e)}"


def check_memory_available() -> str:
    """Check if sufficient memory is available."""
    try:
        memory = psutil.virtual_memory()
        # Alert if less than 10% memory available
        if memory.percent > 90:
            return "degraded: low memory"
        return "healthy"
    except Exception as e:
        return f"unhealthy: {str(e)}"


async def check_database_connection() -> str:
    """Check database connectivity."""
    try:
        # In production, this would check actual database connection
        return "healthy"
    except Exception as e:
        return f"unhealthy: {str(e)}"

File: src/api/routes/health.js
/**
 * Health Check Routes
 * Endpoints for monitoring and health checks
 */

const express = require('express');
const router = express.Router();

// In-memory metrics (would be replaced with actual metrics in production)
const startTime = Date.now();
let requestCount = 0;

/**
 * @route GET /health
 * @desc Basic health check endpoint
 * @access Public
 */
router.get('/', (req, res) => {
  requestCount++;
  res.status(200).json({
    status: 'healthy',
    timestamp: new Date().toISOString(),
    uptime: Math.floor((Date.now() - startTime) / 1000)
  });
});

/**
 * @route GET /health/ready
 * @desc Readiness probe - checks if app can serve traffic
 * @access Public
 */
router.get('/ready', async (req, res) => {
  requestCount++;
  try {
    // Check database connection
    const dbHealthy = await checkDatabaseConnection();

    // Check Redis connection
    const redisHealthy = await checkRedisConnection();

    // Check ML service
    const mlHealthy = await checkMLService();

    const isReady = dbHealthy && redisHealthy;

    res.status(isReady ? 200 : 503).json({
      status: isReady ? 'ready' : 'not_ready',
      timestamp: new Date().toISOString(),
      checks: {
        database: dbHealthy ? 'healthy' : 'unhealthy',
        redis: redisHealthy ? 'healthy' : 'unhealthy',
        ml_service: mlHealthy ? 'healthy' : 'degraded'
      }
    });
  } catch (error) {
    res.status(503).json({
      status: 'not_ready',
      timestamp: new Date().toISOString(),
      error: error.message
    });
  }
});

/**
 * @route GET /health/live
 * @desc Liveness probe - checks if app is running
 * @access Public
 */
router.get('/live', (req, res) => {
  requestCount++;
  res.status(200).json({
    status: 'alive',
    timestamp: new Date().toISOString()
  });
});

/**
 * @route GET /health/metrics
 * @desc Prometheus-compatible metrics endpoint
 * @access Public (should be restricted in production)
 */
router.get('/metrics', async (req, res) => {
  const uptime = Math.floor((Date.now() - startTime) / 1000);
  const memoryUsage = process.memoryUsage();

  const metrics = `
# HELP meal_assistant_uptime_seconds Time since application started
# TYPE meal_assistant_uptime_seconds gauge
meal_assistant_uptime_seconds ${uptime}

# HELP meal_assistant_request_total Total number of health check requests
# TYPE meal_assistant_request_total counter
meal_assistant_request_total ${requestCount}

# HELP meal_assistant_memory_heap_used_bytes Memory heap used in bytes
# TYPE meal_assistant_memory_heap_used_bytes gauge
meal_assistant_memory_heap_used_bytes ${memoryUsage.heapUsed}

# HELP meal_assistant_memory_heap_total_bytes Memory heap total in bytes
# TYPE meal_assistant_memory_heap_total_bytes gauge
meal_assistant_memory_heap_total_bytes ${memoryUsage.heapTotal}

# HELP meal_assistant_memory_rss_bytes Resident set size in bytes
# TYPE meal_assistant_memory_rss_bytes gauge
meal_assistant_memory_rss_bytes ${memoryUsage.rss}

# HELP meal_assistant_memory_external_bytes External memory in bytes
# TYPE meal_assistant_memory_external_bytes gauge
meal_assistant_memory_external_bytes ${memoryUsage.external}

# HELP nodejs_version_info Node.js version
# TYPE nodejs_version_info gauge
nodejs_version_info{version="${process.version}"} 1
`.trim();

  res.set('Content-Type', 'text/plain; charset=utf-8');
  res.send(metrics);
});

/**
 * @route GET /health/info
 * @desc Application info endpoint
 * @access Public
 */
router.get('/info', (req, res) => {
  requestCount++;
  const packageJson = require('../../../package.json');

  res.json({
    name: packageJson.name,
    version: packageJson.version,
    description: packageJson.description,
    environment: process.env.NODE_ENV || 'development',
    nodeVersion: process.version,
    timestamp: new Date().toISOString()
  });
});

// =============================================================================
// Health Check Helper Functions
// =============================================================================

async function checkDatabaseConnection() {
  try {
    // In production, this would check actual database connection
    // Example: await db.query('SELECT 1');
    return true;
  } catch (error) {
    console.error('Database health check failed:', error.message);
    return false;
  }
}

async function checkRedisConnection() {
  try {
    // In production, this would check actual Redis connection
    // Example: await redis.ping();
    return true;
  } catch (error) {
    console.error('Redis health check failed:', error.message);
    return false;
  }
}

async function checkMLService() {
  try {
    // In production, this would check ML service availability
    // Example: await fetch(process.env.ML_SERVICE_URL + '/health');
    const mlServiceUrl = process.env.ML_SERVICE_URL || 'http://localhost:8000';
    // For now, return true as ML service is optional
    return true;
  } catch (error) {
    console.error('ML service health check failed:', error.message);
    return false;
  }
}

module.exports = router;
